dev-docs/RFCs/v5.0/multi-viewport-rfc.md

```diff
@@ -0,0 +1,201 @@
+
+A downside of putting x,y in the Viewport is that it becomes tricky to use the same viewport several times in the list. One could get around this by letting x,y in the viewport descriptor override the viewport but this risks making the system too complicated, and the use case is very limited.
+
+This means that viewport without backing components or other special features will not need a descriptor.
+
+The biggest question is which coordinate system to use for x,y. gl.viewport uses bottom-left and normal CSS layout uses top left. Translating between the two is surprisingly fiddly since both y coordinates and heights need to be stacked (and there could be a DPR issue as well).
+
+The proposal here is to use top left (CSS) coordinates and document it carefully. It is likely to be more natural as the rest of the UI layout with other HTML overlays is done in this coordinate system.
+
+### Viewport Descriptors
+
+## Proposal (Phase 2a): `DeckGL.viewport` prop can take "Viewport Descriptors"
+
+By allowing the `viewport` prop to take "viewport descriptor" objects (in addition) we can address use cases beyond simply rendering multiple views:
+* Multiple render passes over the same viewport - avoid clearing background
+* Rendering to offscreen framebuffers
+
+<DeckGL viewports=[
+  {viewport: new Viewport(...), framebuffer: ...,  clear: ..., parameters: {blendMode, ...}},
+  ...
+]/>
+
+* Alternative: Of course all these options could also be jammed into the base `Viewport` class, but that would add a lot of complexity to an already quite complex class, so allowing separate Viewport descriptors is preferred.
+
+
+### Performance Concerns Around Layer.updateState
+
+updateState({changeFlags: viewportChanged})
+
+The Layer life cycle supports updateState being called when a viewport changes (in addition to when props or data change), so that a layer can do JavaScript calculations using the updated viewport.
+
+When rendering with many viewports there is a concern that updateState gets called many times per frame (potentially recalculating other things that have nothing to do with viewport updates, in less strictly coded layers).  It feels like a perf hit for an edge scenario (though of course still cheaper than maintaining two deck.gl instances in this case).
+
+An important but little known mitigation is that since viewports change very frequently, but few layers actually need to handle this, shouldUpdateState was subtly changed in deck.gl 4.1. It now only returns true when props or data changed, but not when viewports change. So layers that want updateState to be called when viewports change (like ScreenGridLayer) now need to redefine shouldUpdateState. This will mean that even though all layer’s will have shouldUpdateState called every viewport every frame, at least only a few will get calls to updateState.
```

Comment:
> but few layers actually need to handle this
Can you please clarify this, isn't every layers needs to re-render it contents when viewport changes? Do you mean, in multi viewport system, with N viewports, when only 1 viewport is changed, each layer would end up rendering for all N viewports?

Reply:
> Can you please clarify this, isn't every layers needs to re-render it contents when viewport changes? 

updating state and rendering are separate operations. A layer can render many times without a state update. The question here is if the layers updateState needs to be called for every viewport.